### PR TITLE
Implement nonzero for large inputs

### LINKED ISSUE #51871: Tensor.nonzero fails on GPU for tensors containing more than INT_MAX elements
## 🐛 Bug

Calling nonzero for a tensor with more than INT_MAX elements leads to an error on GPU. On CPU there does not seem to be a problem.

## To Reproduce

```python-console
>>> x = torch.zeros(2**20, 2**11, device="cuda")
>>> x.nonzero()
Traceback (most recent call last):
  File "<input>", line 1, in <module>
RuntimeError: nonzero is not supported for tensors with more than INT_MAX elements,   file a support request
```
## Expected behavior

Return the indices of non-zero elements.

## Environment

```
Collecting environment information...
PyTorch version: 1.7.1
Is debug build: False
CUDA used to build PyTorch: 10.2
ROCM used to build PyTorch: N/A

OS: Ubuntu 20.04.2 LTS (x86_64)
GCC version: (Ubuntu 9.3.0-17ubuntu1~20.04) 9.3.0
Clang version: Could not collect
CMake version: version 3.16.3

Python version: 3.8 (64-bit runtime)
Is CUDA available: True
CUDA runtime version: 10.1.243
GPU models and configuration: 
GPU 0: GeForce RTX 2080 Ti
GPU 1: GeForce RTX 2080 Ti
GPU 2: GeForce RTX 2080 Ti
GPU 3: GeForce RTX 2080 Ti

Nvidia driver version: 450.102.04
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip] Could not collect
[conda] Could not collect
```

## Additional context

The problem was already described on [discuss.pytorch.org](https://discuss.pytorch.org/t/torch-geometric/106975), but I could not find the corresponding issue.

cc @ngimel

### PR DESCRIPTION
**Pull request by ngimel**
Fixes #51871

### CODE REVIEW

File: aten/src/ATen/native/cuda/Nonzero.cu [hunk 1]
```diff
@@ -55,86 +53,167 @@ __global__ void write_indices(
...
+      MAX_DIMS,
+      " dimensions");
+  AT_DISPATCH_ALL_TYPES_AND_COMPLEX_AND4(
+      at::ScalarType::ComplexHalf,
+      at::ScalarType::Bool,
+      at::ScalarType::BFloat16,
+      at::ScalarType::Half,
+      self.scalar_type(),
+      "nonzero_cuda",
+      [&] { nonzero_cuda_out_impl<scalar_t>(self, out); });
```
Comment (ezyang) [inline aten/src/ATen/native/cuda/Nonzero.cu:211]: whoops reformatting lol

File: aten/src/ATen/native/cuda/Nonzero.cu [hunk 2]
```diff
@@ -55,86 +53,167 @@ __global__ void write_indices(
...
   auto& allocator = *c10::cuda::CUDACachingAllocator::get();
-  auto num_nonzeros = allocator.allocate(sizeof(int));
-  cub::TransformInputIterator<bool, NonZeroOp<scalar_t>, const scalar_t*> itr(self_.const_data_ptr<scalar_t>(), NonZeroOp<scalar_t>());
-  cub::DeviceReduce::Sum(nullptr, temp_storage_bytes, itr, (int*)num_nonzeros.get(), N, stream);
-  auto temp_storage = allocator.allocate(temp_storage_bytes);
-  cub::DeviceReduce::Sum(temp_storage.get(), temp_storage_bytes, itr, (int*)num_nonzeros.get(), N, stream);
-  int num_nonzeros_h;
+  auto num_nonzeros = allocator.allocate(sizeof(int) * num_chunks);
+  for (int64_t idx = 0; idx < num_chunks; idx++) {
+    int64_t remaining = std::min(chunk_size, self.numel() - idx * chunk_size);
```
Comment (ezyang) [inline aten/src/ATen/native/cuda/Nonzero.cu:75]: yup

File: aten/src/ATen/native/cuda/Nonzero.cu [hunk 3]
```diff
@@ -55,86 +53,167 @@ __global__ void write_indices(
...
+      true, /* pin_memory */
+      std::nullopt /* memory format */
+  );
+  at::cuda::memcpy_and_sync(
+      (void*)pinned_num_nonzeros_h.const_data_ptr<int>(),
+      num_nonzeros.get(),
+      sizeof(int) * num_chunks,
+      cudaMemcpyDeviceToHost,
+      stream);
+  int64_t num_nonzeros_h = 0;
```
Comment (ezyang) [inline aten/src/ATen/native/cuda/Nonzero.cu:109]: yup

**Review by ezyang — APPROVED** [review]

**Comment by eqy** [comment]
did this expose a bug in the rocm allocator???
```
torch.OutOfMemoryError: HIP out of memory. Tried to allocate 67174464.06 GiB.
```
might want to just skip on rocm for now...

**Comment by ngimel** [comment]
Yeah looks like rocm cub is doing weird things. I didn't change the actual cub calls, so the bug was probably always there, skipping the test for now.

**Comment by bhack** [comment]
The upstream team has spent a lot of time benchmarking this with large index. Do you want to add a to-do referencing this so that on new CCCL releases we could benefit the upstream implementation?

**Comment by ngimel** [comment]
Actually no, too little too late from upstream, and adding something that depends on new version leads to loads of #ifdefs that hurt maintainability. If you can show that nonzero is a bottleneck and that upstream implementation is actually faster then we could reconsider.

**Pull request closed without merging**